### PR TITLE
Enhance vmem allocator and slab cache

### PR DESCRIPTION
**Pull request by BrainSlayer**
this patch introduces several new features.
1. implementation of a kvmalloc like interface (kvmem_alloc/kvmem_zalloc)

2. fixes vmem_alloc implementation which can lead to slow allocations
since the first attempt with kmalloc does not make use of the noretry
flag but tells the linux kernel to retry several times before it fails.

3. implements KMC_KVMEM flag for kmem_cache_create which makes use of
a kvmalloc allocator implementation which may increase performance if
the allocator is able to allocate physical memory. otherwise it will
still make use of virtual memory.

4. make use of __GFP_HIGHMEM flag in vmem_alloc , which is required for
some 32 bit systems to make use of full available memory.

5. implement KM_ONCE flag which is behaving almost like KM_SLEEP but
does not hang in a deadloop if no memory is available. it should be used
if you want a GFP_KERNEL like memory allocation, but not that lazy like
KM_NOSLEEP which makes use of GFP_ATOMIC which is mainly made to be used
in irqcontexts. at some locations in zfs like compression algorithms
for instance GFP_ATOMIC/KM_NOSLEEP should not  be used since it might
always fail if blocks are greater than 64-128kb (workspaces). but we also
want no deadlock situation if memory is under pressure. so we need a way
that it behaves like straight GFP_KERNEL. this flag does allow this behaviour.

6. change vmem_alloc to kvmem_alloc on some usefull locations where
memory is initialized.

7. make use of KMC_KVMEM for spl_zlib_workspace_cache, ddt_cache,
zio_buf and zio_data_buf

see also: https://lwn.net/Articles/711653/

Signed-off-by: Sebastian Gottschall <s.gottschall@dd-wrt.com>

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by sempervictus** [comment]
First pass over that actually gave me a mild headache, one of those cases where reading the diff is pretty worthless as it lacks context.
In terms of the retry behavior, has there been any testing to see how it behaves under fragmentation pressure where it has to keep falling back to vmem_alloc? Wondering if there are any "zfs-ish" deadlocks or hangs hiding in the code which this might expose.
I've not built a master branch for some time, but this seems worth it. Hopefully this weekend...
Thank you for the PR.

**Comment by BrainSlayer** [comment]
i have done stress tests with memory constrained systems (512 mb ram, 
slow cpu etc.)
what i have found is that the current kmem_alloc implementation can 
cause deadlocks. but this is unrelated to my patch.
i can reproduce this deadlock also without my patch. it basicly is 
running into a problem if there is no memory available anymore. no 
physical and no virtual.
so kmem_alloc will loop forever. on the other hand. if a internal 
function demands memory and its not available the FS is not function 
anymore anyway.
this deadlock is also not critical since system remains responsive, it 
may applications hang which are using the fs.

my patch does not change much here. the kmem_alloc with KM_VMEM hasnt 
been changed, except that i simplified the code a little bit.
if vmem is requested, it will request  physical memory first but noretry 
flag instead of standard kmalloc. so allocation will be quicker if 
kmalloc fails
since otherwise kmalloc will cause a significant overhead if not enough 
physical memory is available. so we fall qicker back to vmalloc which 
was requested anyway
so code locations which do allocations very often, are getting a benefit 
here since allocation is faster

in addition the new mode KM_KVMEM has been implemented which does ignore 
the kmem_alloc_max boundary but always tries to allocate physical memory 
and falls back
to vmalloc if failed. this has also another performance benefit. virtual 
memory is allocated on first use and has a significant overhead due 
paging. so if possible physical memory should be always
prefered. i converted some static initialized memory allocations from 
VMEM to KVMEM.
KVMEM also does not run into a deadlocking loop since it may fail under 
memory pressure. it should be used with error handling.
in my point of view, if used it forces developers to think about error 
handling instead of using a barrier allocator which may wait forever if 
memory is running out.
so far i only converted uncritical static initializers at driver init to 
KVMEM.

initially i wrote this enhancement on top of the zstd compression 
support since zstd may allocate alot of memory for compression workspace 
and i found out that kmalloc is performing much better than vmalloc
even at bigger sizes and while doing research i found out that kvmalloc 
which was introduced in linux 4.12 is getting about heavily used on alot 
of kernel drivers as replacement for vmalloc.
kvmalloc is faster than vmalloc in case physical memory is available. in 
case physical memory isnt available, its fast as vmalloc before which no 
significant overhead.
read https://lwn.net/Articles/711653/ if you want to know more about the 
intention and behaviour of this function. its basicly the better way 
instead of doing
mem = kmalloc()
if (!mem)
     mem = vmalloc()

since kmalloc is a blocking call and might wait for memory getting 
available for a unspecified retry time. i converted also kmem_cache to 
make use of this new behaviour which may bringt also performance 
benefits to the spl's own slab cache since these allocations are used 
very frequent over running time. so its all about allocation latency and 
memory access performance at all


Am 08.02.2019 um 03:22 schrieb RageLtMan:

**Comment by c0d3z3r0** [comment]
Successfully tested on Debian testing with upstream kernel (4.20.9 - 4.20.17); I have this merged for some weeks and did not have any problems.

**Comment by BrainSlayer** [comment]
i will rework this for current git revision today since there is a small easy to resolve conflict. but thanks for testing

**Comment by c0d3z3r0** [comment]
@BrainSlayer https://github.com/c0d3z3r0/zfs/tree/rebase/pull/8384_vmem

**Comment by BrainSlayer** [comment]
i did this already in my zstd tree which has the same patch

**Comment by c0d3z3r0** [comment]
You should do a rebase on master instead of merging it because you don‘t want the merge commits to get pulled

**Comment by BrainSlayer** [comment]
never did this. git is way to complex and cryptic to be my favorite 
tool. the merge works too anyway

Am 24.03.2019 um 13:54 schrieb Michael Niewöhner:

**Comment by c0d3z3r0** [comment]
#8382

**Comment by c0d3z3r0** [comment]
This improves performance a bit and at least for me did not introduce any problems so far on my machine. Anyone willing to review this? @kpande @behlendorf

**Comment by behlendorf** [comment]
Sorry about the sluggish review.  The thrust of these changes look good, but I'd like to hold of fully reviewing and integrating them until after 0.8 is tagged.  After 0.8 is branched we can get these changes some soak time in master, and backport them if needed to 0.8.x.

**Comment by BrainSlayer** [comment]
consider that this patch is also part of the zstd patch. its a spinoff 
of this project since zstd may allocate much bigger blocks for 
compression workspace which has alot of benefits
of this allocator modifcation.
but please think about the missing GFP_HIGHMEM flag in 
spl_kmem_alloc_impl. from my point of view this is a bug
and may cause uneccessary mem alloc fails on 32 bit systems. (the 
highmem boundary for some mips32 systems may be just 128 mb and is 
variable on x86 depending on the kernel config)
so i think that this small fix should be merged right now into the 
mainline and maybe also backported to 0.7


Am 11.04.2019 um 19:28 schrieb Brian Behlendorf:

**Comment by behlendorf** [comment]
@BrainSlayer can you open a PR with just the minimal `GFP_HIGHMEM` fix so it can be reviewed independently and considered for inclusion.

**Comment by BrainSlayer** [comment]
grrr. another fork of a fork of a fork :-)

Am 11.04.2019 um 19:53 schrieb Brian Behlendorf:

**Comment by c0d3z3r0** [comment]
> grrr. another fork of a fork of a fork :-) Am 11.04.2019 um 19:53 schrieb Brian Behlendorf:

You don't need to fork. This is what branches are made for ;)

**Comment by BrainSlayer** [comment]
then its just a branch for a 8 characters patch which takes a long time 
to setup due the limited capablities of github. consider that my master 
tree is the zstd zfs tree and not the
zfs master. so creating a new branch from a different base is a little 
bit quirky

Am 11.04.2019 um 20:18 schrieb Michael Niewöhner:

**Comment by c0d3z3r0** [comment]
> then its just a branch for a 8 characters patch which takes a long time to setup due the limited capablities of github. consider that my master tree is the zstd zfs tree and not the zfs master. so creating a new branch from a different base is a little bit quirky

Uhm. What? You need to branch off of `master`, then apply your patch, commit, push, pr.
Leave this PR as-is. After the second PR got merged, do a `git rebase origin/master` in this PR's branch.

Example:

~~~sh
git fetch  # update

git branch gfp_highmen origin/master
git checkout gfp_highmen
vim ....   # do your changes
git commit <whatever>
git push -u origin

# wait four gfp_highmem to get merged

git fetch  # update again
git checkout vmem_alloc
git rebase origin/master
git push -f origin vmem_alloc
~~~

**Comment by sempervictus** [comment]
So it boots, upgrades 0.7 test pools in VM, but causes hard lockups in the VM after a while in zloop (no stack trace even to ttyS0), and might also be leaking some kernel addresses via:
```
[Mon Jul  8 07:22:24 2019]  [<ffffffff819c1e0a>] ? pointer+0x47a/0x650
[Mon Jul  8 07:22:24 2019]  [<ffffffff819c222c>] ? vsnprintf+0x24c/0x5d0
[Mon Jul  8 07:22:24 2019]  [<ffffffff8138027e>] ? __dprintf+0xfe/0x160
[Mon Jul  8 07:22:24 2019]  [<ffffffff81252c7b>] ? spl_kmem_zalloc+0x9b/0x130
[Mon Jul  8 07:22:24 2019]  [<ffffffff8131d571>] ? metaslab_sync+0x991/0xba0
[Mon Jul  8 07:22:24 2019]  [<ffffffff81252c00>] ? spl_kmem_zalloc+0x20/0x130
[Mon Jul  8 07:22:24 2019]  [<ffffffff8134d05a>] ? vdev_sync+0x6a/0x1d0
[Mon Jul  8 07:22:24 2019]  [<ffffffff8132fa45>] ? spa_sync+0x5f5/0xf20
[Mon Jul  8 07:22:24 2019]  [<ffffffff8134402e>] ? spa_txg_history_init_io+0xfe/0x110
[Mon Jul  8 07:22:24 2019]  [<ffffffff81347885>] ? txg_sync_thread+0x2b5/0x480
[Mon Jul  8 07:22:24 2019]  [<ffffffff813475d0>] ? txg_thread_exit.isra.0+0xa0/0xa0
[Mon Jul  8 07:22:24 2019]  [<ffffffff8125a67d>] ? thread_generic_wrapper+0x6d/0x80
[Mon Jul  8 07:22:24 2019]  [<ffffffff8125a610>] ? __thread_exit+0x10/0x10
[Mon Jul  8 07:22:24 2019]  [<ffffffff81088fe3>] ? kthread+0x113/0x130
[Mon Jul  8 07:22:24 2019]  [<ffffffff81088ed0>] ? kthread_park+0x80/0x80
[Mon Jul  8 07:22:24 2019]  [<ffffffff8100136c>] ? ret_from_fork+0x2c/0x40
[Mon Jul  8 07:22:29 2019] audit: type=1131 audit(1562570550.153:49): pid=1 ui
```
(automated catch in a grsec kernel - all credit for finding that goes to spender/paxteam).

**Comment by BrainSlayer** [comment]
the stack trace you posted shows that its not caused by this patch. 
spl_kmem_zalloc is a unchanged function path. this patch here changes 
only vmem allocations.
consider also that this patch is tested in productive environment for 
many months together with the zstd patch even with hard load. never seen 
any lockups.
but of course lockups may always occur if the system has no memory left 
for any actions. and i mean really "no memory"

Am 09.07.2019 um 03:19 schrieb RageLtMan:

**Comment by c0d3z3r0** [comment]
@BrainSlayer could you please create another PR for the GFP_HIGHMEM stuff as suggested by @behlendorf, so we can (try to) get stuff merged one-by-one

**Comment by BrainSlayer** [comment]
Am 09.07.2019 um 22:06 schrieb Michael Niewöhner:
>
> @BrainSlayer <https://github.com/BrainSlayer> could you please create 
> another PR for the GFP_HIGHMEM stuff as suggested by @behlendorf 
> <https://github.com/behlendorf>, so we can (try to) get stuff merged 
> one-by-one
>
pff. thats a one liner. need to setup a new git tree first for that patch

**Comment by c0d3z3r0** [comment]
No... just a branch ;) it only takes 1 minute of your time :)

**Comment by c0d3z3r0** [comment]
@BrainSlayer are you sure the highmem flag is really still needed? just found this one... https://github.com/torvalds/linux/commit/19809c2da28aee5860ad9a2eff760730a0710df0

**Comment by BrainSlayer** [comment]
i assume no. for newer kernel revisions. but that doesnt matter we 
should ignore old kernels. whats the minimum kernel requirement?

Am 12.07.2019 um 19:17 schrieb Michael Niewöhner:

**Comment by c0d3z3r0** [comment]
Minimum is 4.12-rc1. You mesnt „should **not** ignore old kernels“, didn‘t you?

**Comment by BrainSlayer** [comment]
4.12-rc1 is clearly not the minimum requirement for zfs. i run it on 
5.2, 4.9, 4.4, 3.18 and even 3.10

Am 13.07.2019 um 05:54 schrieb Michael Niewöhner:

**Comment by c0d3z3r0** [comment]
Yep, then we should include it.. thanks ;)

File: include/spl/sys/kmem.h [hunk 1]
```diff
@@ -41,10 +47,12 @@ extern void strfree(char *str);
 #define	KM_SLEEP	0x0000	/* can block for memory; success guaranteed */
 #define	KM_NOSLEEP	0x0001	/* cannot block for memory; may fail */
 #define	KM_PUSHPAGE	0x0004	/* can block for memory; may use reserve */
+#define	KM_ONCE		0x0008	/* like KM_SLEEP but do not loop allocation */
```
Comment (c0d3z3r0) [inline include/spl/sys/kmem.h]: Where does that get used? I couldn't find any usage here or in your zstd branch

Comment (c0d3z3r0) [reply include/spl/sys/kmem.h]: oops... should have been a comment, not a review....

**Comment by BrainSlayer** [comment]
spl-kmem.c

its not used yet. but may be used. allocating with KM_SLEEP is common in 
zfs, but it has no error handling. under low memory conditions
KM_SLEEP may end in a deadlock / endless loop. situation. using KM_ONCE 
does allow to implement error handling at various locations in zfs if 
required.
so its a usefull api enhancement for future development

Am 13.07.2019 um 07:35 schrieb Michael Niewöhner:

**Comment by BrainSlayer** [comment]
dont bother. you're likely living in the same country as me and it was 
early morning

Am 13.07.2019 um 07:36 schrieb Michael Niewöhner:

**Comment by c0d3z3r0** [comment]
> spl-kmem.c its not used yet. but may be used. allocating with KM_SLEEP is common in zfs, but it has no error handling. under low memory conditions KM_SLEEP may end in a deadlock / endless loop. situation. using KM_ONCE does allow to implement error handling at various locations in zfs if required. so its a usefull api enhancement for future development

Well then, I assume, it should not a) be part of this PR and b) do not get merged until there is any useful implementation. @behlendorf ?

> dont bother. you're likely living in the same country as me and it was early morning

Indeed :D

>>@BrainSlayer could you please create another PR for the GFP_HIGHMEM stuff as suggested by @behlendorf, so we can (try to) get stuff merged one-by-one
>
>pff. thats a one liner. need to setup a new git tree first for that patch

Done now in #9031

**Comment by BrainSlayer** [comment]
thanks for the PR :-) (it wont get merged anyway like all other pull 
requests)

Am 13.07.2019 um 07:53 schrieb Michael Niewöhner:

**Comment by Ornias1993** [comment]
@BrainSlayer I guess this PR requires some edits/updates?
It seems some features are split off into different PR's and It seems at least your intro text doesn't reflect that... Or am I missing something?

**Comment by c0d3z3r0** [comment]
> @BrainSlayer I guess this PR requires some edits/updates?
> It seems some features are split off into different PR's and It seems at least your intro text doesn't reflect that... Or am I missing something?

this is superseeded by my fork; we decided to not integrate KM_ONCE now but when it's needed later; the rest is integrated into my PR

**Comment by Ornias1993** [comment]
@c0d3z3r0 I guess this can be closed then? Or replaced by an issue/todo for later...

**Comment by c0d3z3r0** [comment]
> @c0d3z3r0 I guess this can be closed then? Or replaced by an issue/todo for later...

well, the code is not lost when closed; when it's needed later it's easy to restore it. I'd say this can be closed for now. @BrainSlayer what do you say?

**Comment by Ornias1993** [comment]
@c0d3z3r0 Thats fine indeed, did so before... Closing archive-only PR's works out fine :)

**Comment by BrainSlayer** [comment]
i still like the test system of zfs for improvements. this gets lost if we close the PR. so i can keep up to date my code and still see if there are new problems

**Comment by Ornias1993** [comment]
@BrainSlayer atleast update your PR explaination accordingly then ;)

**Comment by BrainSlayer** [comment]
@Ornias1993 ah now i get it. yes this PR here is dead and unmaintained. i focus on the zstd patch and @c0d3z3r0 took over this part and has his own PR based on my work. i did not check which PR this here was about :-)

**Comment by c0d3z3r0** [comment]
@BrainSlayer then please click "Close"

**Comment by Ornias1993** [comment]
@BrainSlayer No problem, tried to be a little more diplomatic about it :P
And what @c0d3z3r0  says...

**Pull request closed without merging**